components/menu/menu.js

```diff
@@ -0,0 +1,354 @@
+import '../icons/icon.js';
+import './menu-item-return.js';
+import { css, html, LitElement } from 'lit-element/lit-element.js';
+import { HierarchicalViewMixin } from '../hierarchical-view/hierarchical-view-mixin.js';
+
+class Menu extends HierarchicalViewMixin(LitElement) {
+
+	static get properties() {
+		return {
+			active: { type: Boolean, reflect: true },
+			label: { type: String },
+			role: { type: String, reflect: true }
+		};
+	}
+
+	static get styles() {
+		return [ super.styles, css`
+			:host {
+				box-sizing: border-box;
+				display: block;
+				min-width: 180px;
+				width: 100%;
+				padding-top: 1px;
+			}
+
+			:host([active]) .d2l-menu-items d2l-menu-item-return[role="menuitem"],
+			:host([active]) .d2l-menu-items ::slotted([role="menuitem"]),
+			:host([active]) .d2l-menu-items ::slotted([role="menuitemcheckbox"]),
+			:host([active]) .d2l-menu-items ::slotted([role="menuitemradio"]) {
+				position: relative;
+			}
+		`];
+	}
+
+	constructor() {
+		super();
+
+		this.role = 'menu';
+
+		this._keyCodes = {
+			DOWN: 40,
+			ENTER: 13,
+			ESCAPE: 27,
+			LEFT: 37,
+			SPACE: 32,
+			RIGHT: 39,
+			UP: 38
+		};
+		this._items = [];
+	}
+
+	firstUpdated() {
+		super.firstUpdated();
+
+		this.addEventListener('d2l-hierarchical-view-show-start', this._onVisibilityChange);
+		this.addEventListener('d2l-hierarchical-view-hide-complete', this._onVisibilityChange);
+		this.addEventListener('d2l-hierarchical-view-show-complete', this._onShowComplete);
+		this.addEventListener('d2l-hierarchical-view-resize', this._onViewResize);
+		this.addEventListener('d2l-menu-item-visibility-change', this._onMenuItemsChanged);
+		this.addEventListener('keydown', this._onKeyDown);
+		this.addEventListener('keypress', this._onKeyPress);
+
+		this._labelChanged();
+
+		if (this.childView) {
+			const items = this.shadowRoot.querySelector('.d2l-menu-items');
+			items.insertBefore(this._createReturnItem(), items.childNodes[0]);
+		}
+
+		this._onMenuItemsChanged();
+	}
+
+	connectedCallback() {
+		super.connectedCallback();
+
+		this.active = this.getActiveView() === this;
+	}
+
+	updated(changedProperties) {
+		changedProperties.forEach((oldValue, propName) => {
+			if (propName === 'label') {
+				this._labelChanged();
+			}
+		});
+	}
+
+	render() {
+		return html`
+			<div class="d2l-menu-items d2l-hierarchical-view-content">
+				<slot></slot>
+			</div>
+		`;
+	}
+
+	focus() {
```

Comment:
Nit: I think this one should be moved up... not sure if our eslint order rules are working completely?